pkg/operator/controllers/azurensg/subnetnsg_test.go

```diff
@@ -0,0 +1,211 @@
+package azurensg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/mock/gomock"
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maofake "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	mock_network "github.com/Azure/ARO-RP/pkg/util/mocks/azureclient/mgmt/network"
+)
+
+var (
+	vnet = azure.Resource{
+		SubscriptionID: "0000000-0000-0000-0000-000000000000",
+		ResourceGroup:  "vnet-rg",
+		ResourceName:   "vnet-name",
+	}
+)
+
+func TestEnsureSubnetNSG(t *testing.T) {
+	r := AzureNSGReconciler{log: utillog.GetLogger()}
+	for _, tt := range []struct {
+		name                string
+		nsgname             string
+		architectureVersion api.ArchitectureVersion
+		expectedValue       string
+		modifySubnet        func(*mgmtnetwork.Subnet)
+		expectedErr         error
+	}{
+		{
+			name:                "mismatched NSG in Architecture v1",
+			nsgname:             "someotherv1-nsg",
+			architectureVersion: api.ArchitectureVersionV1,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-node-nsg",
+		},
+		{
+			name:                "correct NSG in Architecture v2",
+			nsgname:             "infraid-nsg",
+			architectureVersion: api.ArchitectureVersionV2,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-nsg",
+		},
+		{
+			name:                "mismatched NSG in Architecture v2",
+			nsgname:             "someotherv2-nsg",
+			architectureVersion: api.ArchitectureVersionV2,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-nsg",
+		},
+		{
+			name:                "missing fields",
+			nsgname:             "someotherv2-nsg",
+			architectureVersion: api.ArchitectureVersionV2,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-nsg",
+			modifySubnet: func(subnet *mgmtnetwork.Subnet) {
+				subnet.SubnetPropertiesFormat = nil
+			},
+			expectedErr: fmt.Errorf("received nil, expected a value in SubnetProperties when trying to Get subnet vnet-name/subnet in resource group vnet-rg"),
```

Comment:
Something is wrong with the test. It should fail on this test case because we have this line:

```go
subnetsClient.EXPECT().CreateOrUpdateAndWait(context.Background(), "vnet-rg", "vnet-name", "subnet", fakeSubnet).
	Return(nil)
```

And we do not expect `CreateOrUpdateAndWait` to be called in this case.

I suspect that this is due to missing `defer controller.Finish()`